File: python/array/max_subarray.py
def max_subarray(nums):
    """
    Find the contiguous subarray with the largest sum
    
    Args:
        nums: List[int] - array of integers
    
    Returns:
        int - maximum sum of any contiguous subarray
    """
    if not nums:
        return 0
        
    max_sum = nums[0]
    current_sum = nums[0]
    
    for num in nums[1:]:
        current_sum = max(num, current_sum + num)
        max_sum = max(max_sum, current_sum)
    
    return max_sum

# Example usage:
if __name__ == "__main__":
    test_nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
    result = max_subarray(test_nums)
    print(f"Maximum subarray sum: {result}")  # Expected output: 6

File: go/array/maxsubarray.go
package array

// MaxSubArray finds the contiguous subarray with the largest sum
func MaxSubArray(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    
    maxSum := nums[0]
    currentSum := nums[0]
    
    for i := 1; i < len(nums); i++ {
        // Either start a new subarray at current element,
        // or continue the previous subarray
        if currentSum+nums[i] > nums[i] {
            currentSum += nums[i]
        } else {
            currentSum = nums[i]
        }
        
        // Update maximum sum if current sum is larger
        if currentSum > maxSum {
            maxSum = currentSum
        }
    }
    
    return maxSum
}